addon/-private/system/relationships/relationships-payloads.js

```diff
@@ -0,0 +1,417 @@
+
+    let key = `${modelName}:${relationshipName}`;
+    if (!this._cache[key] && init) {
+      return this._initializeRelationshipPayloads(modelName, relationshipName, modelClass, relationshipsByName);
+    }
+
+    return this._cache[key];
+  }
+
+  _initializeRelationshipPayloads(modelName, relationshipName, modelClass, relationshipsByName) {
+    let relationshipMeta = relationshipsByName.get(relationshipName);
+    let inverseMeta = modelClass.inverseFor(relationshipName, this._store);
+
+    let inverseModelName;
+    let inverseRelationshipName;
+    let inverseRelationshipMeta;
+
+    if (inverseMeta) {
+      inverseRelationshipName = inverseMeta.name
+      inverseModelName = relationshipMeta.type;
+      inverseRelationshipMeta = get(inverseMeta.type, 'relationshipsByName').get(inverseRelationshipName);
+    } else {
+      inverseModelName = inverseRelationshipName = '';
+      inverseRelationshipMeta = null;
+    }
+
+    let lhsKey = `${modelName}:${relationshipName}`;
+    let rhsKey = `${inverseModelName}:${inverseRelationshipName}`;
+
+    return this._cache[lhsKey] =
+      this._cache[rhsKey] =
+      new RelationshipPayloads(
+        this._store,
+        modelName,
+        relationshipName,
+        relationshipMeta,
+        inverseModelName,
+        inverseRelationshipName,
+        inverseRelationshipMeta
+      );
+  }
+}
+
+
+/**
+  Manages the payloads for both sides of a single relationship, across all model
+  instances.
+
+  For example, with
+
+    const User = DS.Model.extend({
+      hobbies: DS.hasMany('hobby')
+    });
+
+    const Hobby = DS.Model.extend({
+      user: DS.belongsTo('user')
+    });
+
+    let relationshipPayloads = new RelationshipPayloads('user', 'hobbies', 'hobby', 'user');
+
+    let userPayload = {
+      data: {
+        id: 1,
+        type: 'user',
+        relationships: {
+          hobbies: {
+            data: [{
+              id: 2,
+              type: 'hobby',
+            }]
+          }
+        }
+      }
+    };
+
+    // here we expect the payload of the individual relationship
+    relationshipPayloads.push('user', 1, 'hobbies', userPayload.data.relationships.hobbies);
+
+    relationshipPayloads.get('user', 1, 'hobbies');
+    relationshipPayloads.get('hobby', 2, 'user');
+
+  @class RelationshipPayloads
+  @private
+*/
+class RelationshipPayloads {
+  constructor(store, modelName, relationshipName, relationshipMeta, inverseModelName, inverseRelationshipName, inverseRelationshipMeta) {
+    this._store = store;
+    this._lhsModelName = modelName;
+    this._lhsRelationshipName = relationshipName;
+    this._lhsRelationshipMeta = relationshipMeta;
+    this._rhsModelName = inverseModelName;
+    this._rhsRelationshipName = inverseRelationshipName;
+    this._rhsRelationshipMeta = inverseRelationshipMeta;
+
+    this._lhsPayloads = {};
+    if (modelName !== inverseModelName || relationshipName !== inverseRelationshipName) {
```

Comment:
I've added some comments in the then clause to explain, as well as expanded the comments in the else clause.

```js
 // then clause

      // The common case of a non-reflexive relationship, or a reflexive
      // relationship whose inverse is not itself

// else clause

      // Edge case when we have a reflexive relationship to itself
      //  eg user hasMany friends inverse friends
      //
      //  In this case there aren't really two sides to the relationship, but
      //  we set `_rhsPayloads = _lhsPayloads` to make things easier to reason
      //  about

```